### ISSUE TITLE
GoogleVR ANE iOS Video not smooth

### ISSUE THREAD

**Issue by Akali65** [myflashlab/GoogleVR-ANE#24]
Before opening a new issue, make sure you are using the latest available ANE. If the ANE is dependent on other libraries or dependencies, also make sure they are the latest versions.

## Environment ##
* iOS Version: 11.0.3 (15A432)
* AIR SDK Version: 26 and 27.0.0.128

## Problem Description ##
Describe your problem in detail.
Using the 2.2.0 Commercial version of the GoogleVR ANE on an iPhone the video seems to be dropping frames and is not playing smoothly when looking around the video scene (up, down, left, right). 

**When I play the same video using the free version of 2.1.1 of GoogleVR ANE the video plays smoothly.**

**Please note** The video on the Android phone plays smoothly when using the 2.2.0 Commercial version of the GoogleVR ANE.

I have followed all the instructions and removed old iOS resources from my app bin folder and copied the new CardboardSDK.bundle, GoogleKitCore.bundle, GoogleKitDialogs.bundle, GoogleKitHUD.bundle, MaterialRobotoFontLoader.bundle from the gvr_iOS_res folder. I've also added all the GoogleToolboxForMac.framework and GTMSessionFetcher.framework to the AIR SDK folder at: AIR_SDK/lib/aot/stub.

* Copy your manifest ```.xml``` file.
```xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<application xmlns="http://ns.adobe.com/air/application/26.0">
    <id>uk.ac.bham</id>
    <versionNumber>1.0</versionNumber>
    <filename>Demo1iOS</filename>
    <description></description>
    <!-- To localize the description, use the following format for the description element.
	<description>
		<text xml:lang="en">English App description goes here</text>
		<text xml:lang="fr">French App description goes here</text>
		<text xml:lang="ja">Japanese App description goes here</text>
	</description>
	-->
    <name>Demo1iOS</name>
    <!-- To localize the name, use the following format for the name element.
	<name>
		<text xml:lang="en">English App name goes here</text>
		<text xml:lang="fr">French App name goes here</text>
		<text xml:lang="ja">Japanese App name goes here</text>
	</name>
	-->
    <copyright></copyright>
    <initialWindow>
        <content>Demo1iOS.swf</content>
        <systemChrome>standard</systemChrome>
        <transparent>false</transparent>
        <visible>true</visible>
        <fullScreen>true</fullScreen>
        <autoOrients>false</autoOrients>
        <renderMode>auto</renderMode>
        <aspectRatio>landscape</aspectRatio>
    </initialWindow>
    <customUpdateUI>false</customUpdateUI>
    <allowBrowserInvocation>false</allowBrowserInvocation>
    <icon>
        <image29x29>AppIconsForPublish/icon_29.png</image29x29>
        <image57x57>AppIconsForPublish/icon_57.png</image57x57>
        <image114x114>AppIconsForPublish/icon_114.png</image114x114>
        <image512x512>AppIconsForPublish/icon_512.png</image512x512>
        <image48x48>AppIconsForPublish/icon_48.png</image48x48>
        <image72x72>AppIconsForPublish/icon_72.png</image72x72>
        <image50x50>AppIconsForPublish/icon_50.png</image50x50>
        <image58x58>AppIconsForPublish/icon_58.png</image58x58>
        <image100x100>AppIconsForPublish/icon_100.png</image100x100>
        <image144x144>AppIconsForPublish/icon_144.png</image144x144>
        <image1024x1024>AppIconsForPublish/icon_1024.png</image1024x1024>
        <image40x40>AppIconsForPublish/icon_40.png</image40x40>
        <image76x76>AppIconsForPublish/icon_76.png</image76x76>
        <image80x80>AppIconsForPublish/icon_80.png</image80x80>
        <image120x120>AppIconsForPublish/icon_120.png</image120x120>
        <image152x152>AppIconsForPublish/icon_152.png</image152x152>
        <image180x180>AppIconsForPublish/icon_180.png</image180x180>
        <image60x60>AppIconsForPublish/icon_60.png</image60x60>
        <image75x75>AppIconsForPublish/icon_75.png</image75x75>
        <image87x87>AppIconsForPublish/icon_87.png</image87x87>
        <image167x167>AppIconsForPublish/icon_167.png</image167x167>
    </icon>
    <supportedLanguages>en</supportedLanguages>
    <iPhone>
        <requestedDisplayResolution>high</requestedDisplayResolution>
        <InfoAdditions><![CDATA[

		<!--iOS 8.0 or higher can support this ANE-->
		<key>MinimumOSVersion</key>
		<string>8.0</string>
		 
		<key>UIPrerenderedIcon</key>
		<true/>
			
		<key>UIStatusBarStyle</key>
		<string>UIStatusBarStyleBlackOpaque</string>
			
		<key>UIRequiresPersistentWiFi</key>
		<string>NO</string>
	
		<key>UIDeviceFamily</key>
		<array><string>1</string></array>

		<!-- Bypass iOS security when debugging your app and you need to load resources from a local server like http://192.168.x.x -->
		<key>NSAppTransportSecurity</key>
		<dict>
			<key>NSAllowsArbitraryLoads</key>
			<true/>
		</dict>
		
]]></InfoAdditions>
    </iPhone>
    <extensions>
        <extensionID>com.myflashlab.air.extensions.googleVR</extensionID>
        <extensionID>com.myflashlab.air.extensions.dependency.overrideAir</extensionID>
    </extensions>
</application>
```

* Copy your Actionscript code.
```actionscript
package 
{
	import com.doitflash.consts.Direction;
	import com.doitflash.consts.Orientation;
	import com.doitflash.mobileProject.commonCpuSrc.DeviceInfo;
	import com.doitflash.starling.utils.list.List;
	import com.doitflash.text.modules.MySprite;
	
	import flash.display.MovieClip;
	import flash.desktop.NativeApplication;
	import flash.desktop.SystemIdleMode;
	import flash.display.Sprite;
	import flash.display.StageAlign;
	import flash.display.StageScaleMode;
	import flash.events.Event;
	import flash.events.InvokeEvent;
	import flash.events.KeyboardEvent;
	import flash.events.MouseEvent;
	import flash.filesystem.File;
	import flash.text.AntiAliasType;
	import flash.text.TextField;
	import flash.text.TextFieldAutoSize;
	import flash.text.TextFormat;
	import flash.text.TextFormatAlign;
	import flash.ui.Keyboard;
	import flash.ui.Multitouch;
	import flash.ui.MultitouchInputMode;
	
	import com.myflashlab.air.extensions.googleVR.*;
	
	public class DemoiOS extends MovieClip
	{
		private const BTN_WIDTH:Number = 150;
		private const BTN_HEIGHT:Number = 60;
		private const BTN_SPACE:Number = 2;
		private var _txt:TextField;
		private var _body:Sprite;
		private var _list:List;
		private var _numRows:int = 1;
		
		public function DemoiOS():void 
		{
			Multitouch.inputMode = MultitouchInputMode.GESTURE;
			NativeApplication.nativeApplication.addEventListener(Event.ACTIVATE, handleActivate);
			NativeApplication.nativeApplication.addEventListener(Event.DEACTIVATE, handleDeactivate);
			NativeApplication.nativeApplication.addEventListener(InvokeEvent.INVOKE, onInvoke);
			NativeApplication.nativeApplication.addEventListener(KeyboardEvent.KEY_DOWN, handleKeys);
			
			stage.addEventListener(Event.RESIZE, onResize);
			stage.scaleMode = StageScaleMode.NO_SCALE;
			stage.align = StageAlign.TOP_LEFT;

			/* Text at the top of the app... Could use this.
			_txt = new TextField();
			_txt.autoSize = TextFieldAutoSize.LEFT;
			_txt.antiAliasType = AntiAliasType.ADVANCED;
			_txt.multiline = true;
			_txt.wordWrap = true;
			_txt.embedFonts = false;
			_txt.htmlText = "<font face='Arimo' color='#333333' size='20'><b>Tarsem Singh Cooner VR2 Test ANE V"+VR.VERSION+"</font>";
			_txt.scaleX = _txt.scaleY = DeviceInfo.dpiScaleMultiplier;
			this.addChild(_txt);
			*/
			
			//adds the buttons at the bottom.
			_body = new Sprite();
			this.addChild(_body);
			
			_list = new List();
			_list.holder = _body;
			_list.itemsHolder = new Sprite();
			_list.orientation = Orientation.VERTICAL;
			_list.hDirection = Direction.LEFT_TO_RIGHT;
			_list.vDirection = Direction.TOP_TO_BOTTOM;
			_list.space = BTN_SPACE;
			
			//Initialises the VR elements.
			init();
		}
		
		private function onInvoke(e:InvokeEvent):void
		{
			NativeApplication.nativeApplication.removeEventListener(InvokeEvent.INVOKE, onInvoke);
		}
		
		private function handleActivate(e:Event):void
		{
			NativeApplication.nativeApplication.systemIdleMode = SystemIdleMode.KEEP_AWAKE;
		}
		
		private function handleDeactivate(e:Event):void
		{
			NativeApplication.nativeApplication.systemIdleMode = SystemIdleMode.NORMAL;
		}
		
		private function handleKeys(e:KeyboardEvent):void
		{
			if(e.keyCode == Keyboard.BACK)
            {
				e.preventDefault();
				NativeApplication.nativeApplication.exit();
            }
		}
		
		private function onResize(e:*=null):void
		{
			if (_txt)
			{
				_txt.width = stage.stageWidth * (1 / DeviceInfo.dpiScaleMultiplier);
			}
			
			if (_list)
			{
				_numRows = Math.floor(stage.stageWidth / (BTN_WIDTH * DeviceInfo.dpiScaleMultiplier + BTN_SPACE));
				_list.row = _numRows;
				_list.itemArrange();
			}
			
			if (_body)
			{
				_body.y = stage.stageHeight - _body.height;
			}
		}
		
		
		private function init():void
		{
			VR.init();
			//Placing this here made no difference at all.
			//VR.api.displayMode = DisplayMode.FULLSCREEN_STEREO;
			
			// add general VR listeners
			VR.api.addEventListener(VREvents.VR_CLOSED, onVRClosed);
			VR.api.addEventListener(VREvents.DID_TAP, onVRTapped);
			VR.api.addEventListener(VREvents.DISPLAY_MODE_CHANGED, onVRModeChanged);
			
			/*----------------------------------------------------------------------
			Removed image stuff from here.
			----------------------------------------------------------------------*/
			
			var btn0:MySprite = createBtn("start video 360");
			btn0.addEventListener(MouseEvent.CLICK, startVid360);
			_list.add(btn0);
			
			function startVid360(e:MouseEvent):void
			{
				VR.api.videoView.removeEventListener(VREvents.LOAD_FAILED, 	onVideoLoadFailed);
				VR.api.videoView.removeEventListener(VREvents.LOAD_SUCCESS, onVideoLoadSuccess);
				
				VR.api.videoView.removeEventListener(VREvents.PAUSED, 		onVideoPaused);
				VR.api.videoView.removeEventListener(VREvents.REACHED_END, 	onVideoReachedEnd);
				VR.api.videoView.removeEventListener(VREvents.RESUMED, 		onVideoResumed);
				VR.api.videoView.removeEventListener(VREvents.STOPPED, 		onVideoStopped);
				VR.api.videoView.removeEventListener(VREvents.POS_UPDATED, 	onVideoPositionUpdated);
				
				/*
				C.log("launching a 360 Video view...");
				trace("On older Android devices you may see your AIR content all black when you return from the 360 video view");
				trace("This is an old bug in AIR and we hope Adobe would find the reason and hopefully fix it someday.");
				trace("In the meanwhile, read this to fix the problem: http://forum.starling-framework.org/topic/ane-fix-for-blackblank-screen-bug-when-returning-to-air-android-apps");
				*/
				var srcVid360:File = File.applicationDirectory.resolvePath("CS1Shorter.mp4");
				var srcExitBtn:File = File.applicationDirectory.resolvePath("exit.png");
				
				var vrSetting:VRConfigVid = new VRConfigVid();
				vrSetting.btnExit = srcExitBtn;
				vrSetting.video360 = srcVid360;
				//vrSetting.video360URL = "http://192.168.0.11/project/products/MyFlashLab_ANEs/googleVR/Vx.x.x/FD/bin/360Vid.mp4";
				//vrSetting.videoType = VRConfig.TYPE_STEREO_OVER_UNDER;
				vrSetting.videoType = VRConfig.TYPE_MONO;
				vrSetting.stereoModeButtonEnabled = true;
				vrSetting.infoButtonEnabled = false;
				vrSetting.touchTrackingEnabled = true;
				vrSetting.fullscreenButtonEnabled = true;
				
				
				VR.api.videoView.addEventListener(VREvents.LOAD_FAILED, 	onVideoLoadFailed);
				VR.api.videoView.addEventListener(VREvents.LOAD_SUCCESS, 	onVideoLoadSuccess);
				
				VR.api.videoView.addEventListener(VREvents.PAUSED, 			onVideoPaused);
				VR.api.videoView.addEventListener(VREvents.REACHED_END, 	onVideoReachedEnd);
				VR.api.videoView.addEventListener(VREvents.RESUMED, 		onVideoResumed);
				VR.api.videoView.addEventListener(VREvents.STOPPED, 		onVideoStopped);
				VR.api.videoView.addEventListener(VREvents.POS_UPDATED, 	onVideoPositionUpdated);
				//If I leave the line below in, the video will not load the fullscreen stereo function.
				//VR.api.displayMode = DisplayMode.FULLSCREEN_STEREO;
				
				VR.api.videoView.attach(vrSetting);
				
				//VR.api.displayMode = DisplayMode.FULLSCREEN_STEREO;
			}
			
			function onVideoLoadFailed(e:VREvents):void
			{
				VR.api.videoView.removeEventListener(VREvents.LOAD_FAILED, onVideoLoadFailed);
				VR.api.videoView.removeEventListener(VREvents.LOAD_SUCCESS, onVideoLoadSuccess);
				
				trace("onVideoLoadFailed: " + e.msg);
			}
			
			function onVideoLoadSuccess(e:VREvents):void
			{
				
				//VR.api.displayMode = DisplayMode.FULLSCREEN_STEREO;
				VR.api.videoView.removeEventListener(VREvents.LOAD_FAILED, onVideoLoadFailed);
				VR.api.videoView.removeEventListener(VREvents.LOAD_SUCCESS, onVideoLoadSuccess);
				trace("onVideoLoadSuccess, video duration = " + e.duration);
			}
			
			function onVideoPaused(e:VREvents):void
			{
				trace("onVideoPaused");
			}
			
			function onVideoResumed(e:VREvents):void
			{
				trace("onVideoResumed");
			}
			
			function onVideoStopped(e:VREvents):void
			{
				trace("onVideoStopped");
			}
			
			function onVideoReachedEnd(e:VREvents):void
			{
				VR.api.videoView.seekTo(0);
				VR.api.videoView.resume();
				
				trace("onVideoReachedEnd");
			}
			
			function onVideoPositionUpdated(e:VREvents):void
			{
				trace("position updated... " + e.position + " / " + e.duration);
			}
			
			onResize();
		}
		
		private function onVRClosed(e:VREvents):void
		{
			trace("vr window closed!");
		}
		
		private function onVRTapped(e:VREvents):void
		{
			trace("vr window tapped!");
			
			trace(VR.api.headRotation);
			
			//VR.api.displayMode = DisplayMode.FULLSCREEN_STEREO;
			
			if (VR.api.videoView.isPaused)
			{
				VR.api.videoView.resume();
			}
			else
			{
				VR.api.videoView.pause();
			}
		}
		
		private function onVRModeChanged(e:VREvents):void
		{
			switch (e.displaymode) 
			{
				case DisplayMode.EMBEDDED:
					trace("vr mode changed: EMBEDDED");
				break;
				
				case DisplayMode.FULLSCREEN_MONO:
					trace("vr mode changed: FULLSCREEN_MONO");
				break;
				
				case DisplayMode.FULLSCREEN_STEREO:
					trace("vr mode changed: FULLSCREEN_STEREO");
				break;
				
				default:
			}
		}	
		
		private function createBtn($str:String):MySprite
		{
			var sp:MySprite = new MySprite();
			sp.addEventListener(MouseEvent.MOUSE_OVER,  onOver);
			sp.addEventListener(MouseEvent.MOUSE_OUT,  onOut);
			sp.addEventListener(MouseEvent.CLICK,  onOut);
			sp.bgAlpha = 1;
			sp.bgColor = 0xDFE4FF;
			sp.drawBg();
			sp.width = BTN_WIDTH * DeviceInfo.dpiScaleMultiplier;
			sp.height = BTN_HEIGHT * DeviceInfo.dpiScaleMultiplier;
			
			function onOver(e:MouseEvent):void
			{
				sp.bgAlpha = 1;
				sp.bgColor = 0xFFDB48;
				sp.drawBg();
			}
			
			function onOut(e:MouseEvent):void
			{
				sp.bgAlpha = 1;
				sp.bgColor = 0xDFE4FF;
				sp.drawBg();
			}
			
			var format:TextFormat = new TextFormat("Arimo", 16, 0x666666, null, null, null, null, null, TextFormatAlign.CENTER);
			
			var txt:TextField = new TextField();
			txt.autoSize = TextFieldAutoSize.LEFT;
			txt.antiAliasType = AntiAliasType.ADVANCED;
			txt.mouseEnabled = false;
			txt.multiline = true;
			txt.wordWrap = true;
			txt.scaleX = txt.scaleY = DeviceInfo.dpiScaleMultiplier;
			txt.width = sp.width * (1 / DeviceInfo.dpiScaleMultiplier);
			txt.defaultTextFormat = format;
			txt.text = $str;
			
			txt.y = sp.height - txt.height >> 1;
			sp.addChild(txt);
			
			return sp;
		}
	}
}
```
## Steps to Reproduce ##
Use the same AIR SDK and commercial version of GoogleVR ANE on an iOS iPhone 6 plus (make sure you have the right size launch image 2208x1242 landscape). 

I have used the following AIR for iOS Settings:
![screen shot 2017-11-04 at 10 24 34](https://user-images.githubusercontent.com/24958799/32404600-785e449a-c14a-11e7-980c-019aab7369b5.png)

The video settings are as follows:
![screen shot 2017-11-04 at 10 26 41](https://user-images.githubusercontent.com/24958799/32404620-d2b0456a-c14a-11e7-8b69-823349682dca.png)

**Comment by myflashlab** [comment]
Considering that almost all our ANEs are somehow linked to each other and besides, their dependencies to different frameworks/resources, we do not support older versions. That said, it's worthy to mention that we are right now working on an updated version of the VR ANE which might help you with your problem. So, wait on the new release.

In general, please consider that our job is to make the native code work on the AIR side and any questions about how the ANE is implemented and compiled, we can answer. But questions about how the SDK works can't be answered with us. you should try to ask such these questions from the original developers of the SDK, in this case, you can ask this question here: https://github.com/googlevr/gvr-ios-sdk

**Comment by Akali65** [comment]
Thank you for your response. Do you have a timeframe for when the update will be available please?

Also, is it possible for me to join any beta testing programme you may have for this ANE?

I love this ANE, thank you for taking the time to produce it. I hope that the update will be available soon and will resolve the problem I am currently having.

**Comment by myflashlab** [comment]
The upgrade is on its way and we are waiting for Adobe's alpha release of AIR 28 as a LOT will be changed with the new AIR SDK. We will announce it when ready. we're on it :)
